File: android/src/main/java/com/onmr/flutter/wechat/WechatPlugin.java
package com.onmr.flutter.wechat;

import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.plugin.common.EventChannel.StreamHandler;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Bitmap.CompressFormat;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.widget.Toast;

import com.tencent.mm.opensdk.modelbase.BaseReq;
import com.tencent.mm.opensdk.modelbase.BaseResp;
import com.tencent.mm.opensdk.modelmsg.SendAuth;
import com.tencent.mm.opensdk.modelmsg.SendMessageToWX;
import com.tencent.mm.opensdk.modelmsg.WXImageObject;
import com.tencent.mm.opensdk.modelmsg.WXMediaMessage;
import com.tencent.mm.opensdk.modelmsg.WXMusicObject;
import com.tencent.mm.opensdk.modelmsg.WXTextObject;
import com.tencent.mm.opensdk.modelmsg.WXVideoObject;
import com.tencent.mm.opensdk.modelmsg.WXWebpageObject;
import com.tencent.mm.opensdk.modelpay.PayReq;
import com.tencent.mm.opensdk.openapi.IWXAPI;
import com.tencent.mm.opensdk.openapi.IWXAPIEventHandler;
import com.tencent.mm.opensdk.openapi.WXAPIFactory;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;

/** WechatPlugin */
public class WechatPlugin implements MethodCallHandler {

  private static int code;//返回错误吗
  private static String loginCode;//获取access_code
  private static IWXAPI api;
  private static Result result;
  private Context context;
  private String appid;
  private Bitmap bitmap;
  private WXMediaMessage message;
  private String kind = "session";
  private final PluginRegistry.Registrar registrar;
  private BroadcastReceiver sendRespReceiver;

  private WechatPlugin(Context ctx, Registrar registrar) {
    this.registrar = registrar;
    context = ctx;
  }

  public static int getCode() {
    return code;
  }

  public static void setCode(int value) {
    code = value;
  }

  public static String getLoginCode() {
    return loginCode;
  }

  public static void setLoginCode(String value) {
    loginCode = value;
  }

  private Handler handler = new Handler(new Handler.Callback() {
    @Override
    public boolean handleMessage(Message osMessage) {
      SendMessageToWX.Req request = new SendMessageToWX.Req();
      request.scene = kind.equals("timeline")
        ? SendMessageToWX.Req.WXSceneTimeline
        : kind.equals("favorite")
          ? SendMessageToWX.Req.WXSceneFavorite
          : SendMessageToWX.Req.WXSceneSession;

      if (bitmap != null) {
        Bitmap thumbBitmap = Bitmap.createScaledBitmap(bitmap, 100, 100, true);
        message.thumbData = convertBitmapToByteArray(thumbBitmap, true);
      }
      switch (osMessage.what) {
        case 0:
          request.transaction = String.valueOf(System.currentTimeMillis());
          request.message = message;
          api.sendReq(request);
          break;
        case 1:
          if (bitmap == null) {
            Toast.makeText(context, "图片路径错误", Toast.LENGTH_SHORT).show();
            break;
          }
          WXImageObject imageObject = new WXImageObject(bitmap);
          message.mediaObject = imageObject;
          request.transaction = String.valueOf(System.currentTimeMillis());
          request.message = message;
          api.sendReq(request);
          break;
        case 2:
          request.transaction = String.valueOf(System.currentTimeMillis());
          request.message = message;
          api.sendReq(request);
          break;
        case 3:
          request.transaction = String.valueOf(System.currentTimeMillis());
          request.message = message;
          api.sendReq(request);
          break;
        default:
          break;
      }
      return false;
    }
  });

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "wechat");
    channel.setMethodCallHandler(new WechatPlugin(registrar.context(), registrar));
    IntentFilter intentFilter = new IntentFilter();
    intentFilter.addAction("sendResp");
    registrar.context().registerReceiver(createReceiver(), intentFilter);
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    }
    else if (call.method.equals("register")) {
      appid = call.argument("appid");
      api = WXAPIFactory.createWXAPI(context, appid, true);
      result.success(api.registerApp(appid));
    }
    // Check if wechat app installed
    else if (call.method.equals("isWechatInstalled")) {
      if (api == null) {
        result.success(false);
      } else {
        result.success(api.isWXAppInstalled());
      }
    }
    else if (call.method.equals("getApiVersion")) {
      result.success(api.getWXAppSupportAPI());
    }
    else if (call.method.equals("openWechat")) {
      result.success(api.openWXApp());
    }
    else if (call.method.equals("share")) {
      final String kind = call.argument("kind");
      final String to = call.argument("to");
      final String coverUrl = call.argument("coverUrl");
      SendMessageToWX.Req request = new SendMessageToWX.Req();
      message = new WXMediaMessage();
      request.scene = to.equals("timeline")
        ? SendMessageToWX.Req.WXSceneTimeline
        : to.equals("favorite")
          ? SendMessageToWX.Req.WXSceneFavorite
          : SendMessageToWX.Req.WXSceneSession;
      switch (kind) {
        case "text":
          WXTextObject textObject = new WXTextObject();
          final String text = call.argument("text");
          textObject.text = text;
          message.mediaObject = textObject;
          message.description = text;
          request.transaction = String.valueOf(System.currentTimeMillis());
          request.message = message;
          api.sendReq(request);
          break;
        case "music":
          WXMusicObject musicObject = new WXMusicObject();
          musicObject.musicUrl = call.argument("resourceUrl").toString();
          musicObject.musicDataUrl = call.argument("resourceUrl").toString();
          musicObject.musicLowBandDataUrl = call.argument("resourceUrl").toString();
          musicObject.musicLowBandUrl = call.argument("resourceUrl").toString();
          message = new WXMediaMessage();
          message.mediaObject = musicObject;
          message.title = call.argument("title").toString();
          message.description = call.argument("description").toString();
          //网络图片或者本地图片
          new Thread() {
            public void run() {
              Message osMessage = new Message();
              bitmap = GetBitmap(coverUrl);
              osMessage.what = 2;
              handler.sendMessage(osMessage);
            }
          }.start();
          break;
        case "video":
          WXVideoObject videoObject = new WXVideoObject();
          videoObject.videoUrl = call.argument("resourceUrl").toString();
          videoObject.videoLowBandUrl = call.argument("resourceUrl").toString();
          message = new WXMediaMessage();
          message.mediaObject = videoObject;
          message.title = call.argument("title").toString();
          message.description = call.argument("description").toString();
          //网络图片或者本地图片
          new Thread() {
            public void run() {
              Message osMessage = new Message();
              bitmap = GetBitmap(coverUrl);
              osMessage.what = 3;
              handler.sendMessage(osMessage);
            }
          }.start();
          break;
        case "image":
          message = new WXMediaMessage();
          message.title = call.argument("title").toString();
          message.description = call.argument("description").toString();
          final String imageResourceUrl = call.argument("resourceUrl");
          //网络图片或者本地图片
          new Thread() {
            public void run() {
              Message osMessage = new Message();
              bitmap = GetBitmap(imageResourceUrl);
              osMessage.what = 1;
              handler.sendMessage(osMessage);
            }
          }.start();
          break;
        case "webpage":
          WXWebpageObject webpageObject = new WXWebpageObject();
          webpageObject.webpageUrl = call.argument("url").toString();
          message = new WXMediaMessage();
          message.mediaObject = webpageObject;
          message.title = call.argument("title").toString();
          message.description = call.argument("description").toString();
          //网络图片或者本地图片
          new Thread() {
            public void run() {
              Message osMessage = new Message();
              bitmap = GetBitmap(coverUrl);
              osMessage.what = 0;
              handler.sendMessage(osMessage);
            }
          }.start();
          break;
      }
    }
    else if (call.method.equals("login")) {
      final String scope = call.argument("scope").toString();
      final String state = call.argument("state").toString();
      SendAuth.Req authRequest = new SendAuth.Req();
      authRequest.scope = scope;
      authRequest.state = state;
      api.sendReq(authRequest);
    }
    else if (call.method.equals("pay")) {
      final String appId = call.argument("appId").toString();
      final String partnerId = call.argument("partnerId").toString();
      final String prepayId = call.argument("prepayId").toString();
      final String nonceStr = call.argument("nonceStr").toString();
      final String timestamp = call.argument("timestamp").toString();
      final String sign = call.argument("sign").toString();
      final String packageValue = call.argument("package").toString();
      PayReq payRequest = new PayReq();
      payRequest.partnerId = partnerId;
      payRequest.prepayId = prepayId;
      payRequest.nonceStr = nonceStr;
      payRequest.timeStamp = timestamp;
      payRequest.sign = sign;
      payRequest.packageValue = packageValue;
      payRequest.appId = appId;
      api.sendReq(payRequest);
    }
    else {
      result.notImplemented();
    }
  }

  public Bitmap GetBitmap(String url) {
    Bitmap bitmap = null;
    InputStream in = null;
    BufferedOutputStream out = null;
    try {
      in = new BufferedInputStream(new URL(url).openStream(), 1024);
      final ByteArrayOutputStream dataStream = new ByteArrayOutputStream();
      out = new BufferedOutputStream(dataStream, 1024);
      copy(in, out);
      out.flush();
      byte[] data = dataStream.toByteArray();
      bitmap = BitmapFactory.decodeByteArray(data, 0, data.length);
      return bitmap;
    } catch (IOException e) {
      e.printStackTrace();
      return null;
    }
  }

  public byte[] convertBitmapToByteArray(final Bitmap bitmap, final boolean needRecycle) {
    ByteArrayOutputStream output = new ByteArrayOutputStream();
		bitmap.compress(CompressFormat.PNG, 100, output);
		if (needRecycle) {
			bitmap.recycle();
		}
		
		byte[] result = output.toByteArray();
		try {
			output.close();
		} catch (Exception e) {
			e.printStackTrace();
		}
		
		return result;
  }

  private static void copy(InputStream in, OutputStream out) throws IOException {
    byte[] b = new byte[1024];
    int read;
    while ((read = in.read(b)) != -1) {
      out.write(b, 0, read);
    }
  }

  private static BroadcastReceiver createReceiver() {
    return new BroadcastReceiver() {
      @Override
      public void onReceive(Context context, Intent intent) {
        System.out.println(intent.getStringExtra("type"));
        if (intent.getStringExtra("type").equals("SendAuthResp")) {
          result.success(intent.getStringExtra("code"));
        }
        else if (intent.getStringExtra("type").equals("PayResp")) {
          result.success(intent.getStringExtra("code"));
        }
        else if (intent.getStringExtra("type").equals("ShareResp")) {
          System.out.println(intent.getStringExtra("code"));
          result.success(intent.getStringExtra("code"));
        }
      }
    };
  }
}

File: ios/Classes/WechatPlugin.m
#import "WechatPlugin.h"

@implementation WechatPlugin {
  FlutterResult result;
}

- (instancetype)init {
  self = [super init];
  [[NSNotificationCenter defaultCenter]
    addObserver:self
    selector:@selector(handleOpenURL:)
    name:@"WeChat"
    object:nil
  ];
  return self;
}

- (void)dealloc {
  [[NSNotificationCenter defaultCenter] removeObserver:self];
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"wechat"
            binaryMessenger:[registrar messenger]];
  WechatPlugin* instance = [[WechatPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSDictionary *arguments = [call arguments];
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  }
  // Register app to Wechat with appid
  else if ([@"register" isEqualToString:call.method]) {
    [WXApi registerApp:arguments[@"appid"]];
    result(nil);
  }

  else if ([@"isWechatInstalled" isEqualToString:call.method]) {
    BOOL installed = [WXApi isWXAppInstalled];
    result([NSString stringWithFormat:@"%@", installed ? @"true" : @"false"]);
  }

  else if ([@"getApiVersion" isEqualToString:call.method]) {
    NSString *apiVersion = [WXApi getApiVersion];
    result([NSString stringWithFormat:@"%@", apiVersion]);
  }

  else if ([@"openWechat" isEqualToString:call.method]) {
    result(nil);
  }

  // Share something to Wechat
  else if ([@"share" isEqualToString:call.method]) {
    
    // What kind of things want to share
    NSString* kind = arguments[@"kind"];
    NSString* to = arguments[@"to"];
    
    // Wechat Message Request.
    SendMessageToWXReq *request = [[SendMessageToWXReq alloc] init];
    // Defaults request is not bText.
    request.bText = NO;
    // Share to wechat timeline.
    if ([@"timeline" isEqualToString:to]) {
      request.scene = WXSceneTimeline;
    }
    // Add to wechat favorites list.
    else if ([@"favorite" isEqualToString:to]) {
      request.scene = WXSceneFavorite;
    }
    // Defaults share to wechat session.
    else {
      request.scene = WXSceneSession;
    }

    // Sharing some text content
    if ([@"text" isEqualToString:kind]) {
      // Text content
      NSString *text = arguments[@"text"];
      request.text = text;
      request.bText = YES;
    }
    // Sharing an image
    else if ([@"image" isEqualToString:kind]) {
      // Image imageUrl
      NSString *resourceUrl = arguments[@"resourceUrl"];

      WXMediaMessage *mediaMessage = [WXMediaMessage message];
      WXImageObject *imageObject = [WXImageObject object];
      imageObject.imageData = [NSData dataWithContentsOfURL:[NSURL URLWithString:resourceUrl]];
      mediaMessage.mediaObject = imageObject;
  
      request.message = mediaMessage;
    }
    // Sharing music
    else if ([@"music" isEqualToString:kind]) {
      // music resource Url
      NSString *resourceUrl = arguments[@"resourceUrl"];
      // title of this music
      NSString *title = arguments[@"title"];
      // description of this music
      NSString *description = arguments[@"description"];
      // webpage url of this music
      NSString *url = arguments[@"url"];
      // cover Url of this music
      NSString *coverUrl = arguments[@"coverUrl"];
      NSData* coverImageData = [NSData dataWithContentsOfURL:[NSURL URLWithString:coverUrl]];

      WXMusicObject *musicObject = [WXMusicObject object];
      musicObject.musicUrl = url;
      musicObject.musicDataUrl = resourceUrl;

      WXMediaMessage *mediaMessage = [WXMediaMessage message];
      mediaMessage.title = title;
      mediaMessage.description = description;
      [mediaMessage setThumbImage:[UIImage imageWithData:coverImageData]];
      mediaMessage.mediaObject = musicObject;

      request.message = mediaMessage;
    }
    // Sharing video
    else if ([@"video" isEqualToString:kind]) {
      NSString *resourceUrl = arguments[@"resourceUrl"];
      NSString *coverUrl = arguments[@"coverUrl"];
      // NSString *url = arguments[@"url"];
      NSString *title = arguments[@"title"];
      NSString *description = arguments[@"description"];

      NSData* coverImageData = [NSData dataWithContentsOfURL:[NSURL URLWithString:coverUrl]];

      WXVideoObject *mediaObject = [WXVideoObject object];
      mediaObject.videoUrl = resourceUrl;

      WXMediaMessage *mediaMessage = [WXMediaMessage message];
      mediaMessage.title = title;
      mediaMessage.description = description;
      mediaMessage.mediaObject = mediaObject;
      [mediaMessage setThumbImage:[UIImage imageWithData:coverImageData]];

      request.message = mediaMessage;
    }
    // Sharing webpage
    else if ([@"webpage" isEqualToString:kind]) {
      NSString *coverUrl = arguments[@"coverUrl"];
      NSString *url = arguments[@"url"];
      NSString *title = arguments[@"title"];
      NSString *description = arguments[@"description"];

      NSData* coverImageData = [NSData dataWithContentsOfURL:[NSURL URLWithString:coverUrl]];

      WXWebpageObject *mediaObject = [WXWebpageObject object];
      mediaObject.webpageUrl = url;

      WXMediaMessage *mediaMessage = [WXMediaMessage message];
      mediaMessage.title = title;
      mediaMessage.description = description;
      mediaMessage.mediaObject = mediaObject;
      [mediaMessage setThumbImage:[UIImage imageWithData:coverImageData]];

      request.message = mediaMessage;
    }
    [WXApi sendReq:request];
  }

  // Login via wechat
  else if ([@"login" isEqualToString:call.method]) {
    NSString* scope= arguments[@"scope"];
    NSString* state= arguments[@"state"];
    SendAuthReq *request = [[SendAuthReq alloc] init];
    request.scope = scope;
    request.state = state;

    [WXApi sendReq:request];
  }

  // Pay via wechat
  else if ([@"pay" isEqualToString:call.method]) {
    NSString *partnerId = arguments[@"partnerId"];
    NSString *prepayId = arguments[@"prepayId"];
    NSString *package = arguments[@"package"];
    NSString *nonceStr = arguments[@"nonceStr"];
    NSString *timestamp= arguments[@"timestamp"];
    NSString *sign= arguments[@"sign"];

    PayReq *request = [[PayReq alloc] init];
    request.partnerId = partnerId;
    request.prepayId = prepayId;
    request.package = package;
    request.nonceStr = nonceStr;
    request.timeStamp = [timestamp intValue];
    request.sign = sign;

    [WXApi sendReq:request];
  }

  else {
    result(FlutterMethodNotImplemented);
  }
}

-(BOOL)handleOpenURL:(NSNotification *)notification {
  NSString * urlString =  [notification userInfo][@"url"];
  NSURL * url = [NSURL URLWithString:urlString];
  if ([WXApi handleOpenURL:url delegate:self]) {
    return YES;
  }
  else {
    return NO;
  }
}

-(void) onResp:(BaseResp*)resp {
  if([resp isKindOfClass:[SendMessageToWXResp class]]) {
    result([NSString stringWithFormat:@"%d",resp.errCode]);
  }
  else if ([resp isKindOfClass:[SendAuthResp class]]) {
    SendAuthResp *r = (SendAuthResp *)resp;
    if (r.errCode == WXSuccess) {
      result(r.code);
    }
    else {
      result([NSString stringWithFormat:@"%d",r.errCode]);
    }
  }
  else if ([resp isKindOfClass:[PayResp class]]) {
    result([NSString stringWithFormat:@"%d",resp.errCode]);
  }
}

@end
